File: lesson-04-linq-collections-functional/samples/L04.QuoteData/Rating.cs
namespace L04.QuoteData;

public sealed record RatingInput(
    CoverageClass Coverage,
    decimal SumInsured,
    int DriverAge,
    int Claims,
    int LicenceYears,
    VehicleUse Use,
    int EngineCc);

/// <summary>The track's one illustrative tariff (_curriculum.md, "Canonical tariff"), not a real one.
/// Every member is a pure function.</summary>
public static class Rating
{
    /// <summary>3 or more claims in 5 years: a decision, not a price. The quote is never rated.</summary>
    public const string DeclineReason = "3+ claims in 5 years";

    public static bool IsDeclined(RatingInput r) => r.Claims >= 3;

    #region rules-as-functions
    // a rule is a value: it reads the quote and returns the loading it adds to the base premium
    public delegate decimal Loading(RatingInput input);

    public static readonly IReadOnlyList<(string Name, Loading Rate)> Rules =
    [
        ("young driver", r => r.DriverAge < 25 ? 0.20m : 0m),
        ("claims", r => r.Claims switch
        {
            0 => 0m, 1 => 0.10m, 2 => 0.25m,
            _ => throw new InvalidOperationException(DeclineReason), // 3+ is declined, not loaded
        }),
        ("commercial use", r => r.Use is not VehicleUse.Commercial ? 0m
                                : r.EngineCc > 3_000 ? 0.35m : 0.25m),
    ];

    // fold the rules into one loading, then apply it and the no-claim discount to the base premium
    public static decimal Net(RatingInput r) =>
        Round(Base(r) * (1m + Rules.Aggregate(0m, (loading, rule) => loading + rule.Rate(r)))
                      * (1m - NoClaimDiscount(r)));

    // claim-free years earn the ladder; a single claim resets it to zero
    static decimal NoClaimDiscount(RatingInput r) =>
        (r.Claims == 0 ? r.LicenceYears : 0) switch
        {
            0 => 0m, 1 => 0.20m, 2 => 0.25m, 3 => 0.30m, 4 => 0.40m, _ => 0.50m,
        };
    #endregion

    public static decimal Base(RatingInput r) => r.Coverage switch
    {
        CoverageClass.Class1 => r.SumInsured * 0.021m,
        CoverageClass.Class2Plus => r.SumInsured * 0.012m,
        CoverageClass.Class3Plus => r.SumInsured * 0.009m,
        CoverageClass.Class3 => r.SumInsured * 0.004m, // third-party only: a rate, never a flat amount
        _ => throw new ArgumentOutOfRangeException(nameof(r)),
    };

    public static Money Total(RatingInput r)
    {
        var net = Net(r);
        var duty = Round(net * 0.004m);        // stamp duty 0.4% of the net premium
        var vat = Round((net + duty) * 0.07m); // VAT 7% on net + stamp duty
        return new Money(net + duty + vat);
    }

    // the satang: two decimal places, halves away from zero
    static decimal Round(decimal amount) => decimal.Round(amount, 2, MidpointRounding.AwayFromZero);

    public static int AgeOn(DateOnly dateOfBirth, DateOnly asOf)
    {
        var age = asOf.Year - dateOfBirth.Year;
        return asOf < dateOfBirth.AddYears(age) ? age - 1 : age;
    }
}

File: lesson-12-cloud-native-modernization/samples/L12.RatingVb/Rating.vb
' MotorQuote rating rules: the one illustrative tariff of the whole track
' (curriculum, "Canonical tariff"). ILLUSTRATIVE rates only — not a real motor tariff.

Public Enum CoverageClass
    Class1       ' comprehensive
    Class2Plus
    Class3Plus
    Class3       ' third-party only
End Enum

''' <summary>Net premium, stamp duty, VAT and total, in THB.</summary>
Public Structure PremiumBreakdown
    Public ReadOnly Property Net As Decimal
    Public ReadOnly Property StampDuty As Decimal
    Public ReadOnly Property Vat As Decimal
    Public ReadOnly Property Total As Decimal

    Public Sub New(net As Decimal, stampDuty As Decimal, vat As Decimal)
        Me.Net = net
        Me.StampDuty = stampDuty
        Me.Vat = vat
        Total = net + stampDuty + vat
    End Sub
End Structure

''' <summary>
''' Three or more claims in five years: the quote is declined. A domain outcome, not a crash,
''' so every host answers it with HTTP 422. (Exceptions are lesson 02's subject, not this one's.)
''' </summary>
Public NotInheritable Class QuoteDeclinedException
    Inherits Exception

    Public Sub New()
        MyBase.New("declined: 3 or more claims in the last 5 years")
    End Sub
End Class

Public Module Rating
#Region "rating"
    ' A Module compiles to a static class: C# calls Rating.Quote(...).
    ' driverAge is the driver's age on the quote start date.
    Public Function Quote(coverage As CoverageClass, sumInsured As Decimal,
                          driverAge As Integer, licenceYears As Integer,
                          claimsLast5Years As Integer, commercial As Boolean,
                          engineCc As Integer) As PremiumBreakdown
        If claimsLast5Years >= 3 Then Throw New QuoteDeclinedException()

        Dim rate As Decimal
        Select Case coverage
            Case CoverageClass.Class1 : rate = 0.021D
            Case CoverageClass.Class2Plus : rate = 0.012D
            Case CoverageClass.Class3Plus : rate = 0.009D
            Case Else : rate = 0.004D
        End Select

        ' loadings add up, then apply to the base premium
        Dim loadings = 0D
        If driverAge < 25 Then loadings += 0.2D
        If claimsLast5Years = 1 Then loadings += 0.1D
        If claimsLast5Years = 2 Then loadings += 0.25D
        If commercial Then loadings += If(engineCc > 3000, 0.35D, 0.25D)

        ' the no-claim ladder counts claim-free licence years
        Dim claimFree = If(claimsLast5Years = 0, licenceYears, 0)
        Dim discount As Decimal
        Select Case claimFree
            Case Is <= 0 : discount = 0D
            Case 1 : discount = 0.2D
            Case 2 : discount = 0.25D
            Case 3 : discount = 0.3D
            Case 4 : discount = 0.4D
            Case Else : discount = 0.5D
        End Select

        Dim net = Satang(sumInsured * rate * (1D + loadings) * (1D - discount))
        Dim duty = Satang(net * 0.004D)
        Dim vat = Satang((net + duty) * 0.07D)
        Return New PremiumBreakdown(net, duty, vat)
    End Function

    ' the satang: 2 decimals, halves round away from zero
    Private Function Satang(amount As Decimal) As Decimal
        Return Math.Round(amount, 2, MidpointRounding.AwayFromZero)
    End Function
#End Region
End Module
